### PR TITLE
Fix line kick and forward pass.

### LINKED ISSUE #264: Replan just the last segment when there is a goal change
Should be able to implement a very simple direct path for small goal changes. Add the plan onto the end of the path.

### PR DESCRIPTION
**Pull request by ashaw596**
This isn't really ready, but it should be pretty close after I get a chance to test with the robots.

TODO

- [ ] Test with real robots
- [ ] Fix up Ball prediction with real ball
- [ ] Clean up everything.

Closes #264

### CODE REVIEW

**Comment by jgkamat** [comment]
@ashaw596 the 'CompositePath.CompositeSubPath' test seems to be failing, not really that urgent (since this is still being worked on), but just giving you a heads up.

Let me know when you think this is ready and I'll look over this. Thanks! We definetly need better lineup (for all the restarts/corner kicks) so this is pretty critical (and we have to get it right).

**Comment by ashaw596** [comment]
Ok. Fixed the tests at least I think.

File: soccer/planning/SingleRobotPathPlanner.cpp [hunk 1]
```diff
@@ -115,10 +118,10 @@ bool SingleRobotPathPlanner::shouldReplan(const PlanRequest& planRequest) {
 
     // if this number of microseconds passes since our last path plan, we
     // automatically replan
-    const RJ::Seconds kPathExpirationInterval = RJ::Seconds(replanTimeout());
-    if ((RJ::now() - prevPath->startTime()) > kPathExpirationInterval) {
-        return true;
-    }
+    //const RJ::Seconds kPathExpirationInterval = RJ::Seconds(replanTimeout());
```
Comment (jgkamat) [inline soccer/planning/SingleRobotPathPlanner.cpp]: Can this be deleted?

File: soccer/planning/CompositePath.cpp [hunk 2]
```diff
@@ -80,13 +89,13 @@ bool CompositePath::hit(const Geometry2d::ShapeSet& obstacles,
     }
 
     for (; start < paths.size(); start++) {
-        if (paths[start]->hit(obstacles, 0ms, hitTime)) {
+        if (paths.at(start)->hit(obstacles, 0ms, hitTime)) {
```
Comment (ashaw596) [inline soccer/planning/CompositePath.cpp]: yeah pretty much. I don't trust myself with out of bounds arrays.

**Comment by jgkamat** [comment]
@ashaw596 do we want to try to get this in before comp? I think that we should, since the line kick is pretty central to our restarts. I want to nail it down as much as possible before comp if we do though.

**Comment by ashaw596** [comment]
Yeah definitely. I'll be back in Atlanta next Friday. I'll start cleaning
it up now and I'll fine tune it some more when I get back.

**Comment by jgkamat** [comment]
@ashaw596 is this ready for a 2nd round of reviews, I'd realllllly like to get this in before comp if only to save us the headache

**Comment by ashaw596** [comment]
Yeah. We're finishing up some motion control tuning/prevent motors from burning out with jeremy rn and I'll test this after the robots are moving. Hopefully tonight!

**Comment by jgkamat** [comment]
blocked on https://github.com/RoboJackets/robocup-software/pull/1001 currently. Once that's merged this can be continued to merge this into master.

File: soccer/gameplay/plays/restarts/our_free_kick.py [hunk 3]
```diff
@@ -38,6 +38,8 @@ def __init__(self, indirect=None):
         # This will be reset to something else if indirect on the first iteration
         kicker.target = constants.Field.TheirGoalSegment
 
+        # kicker.set_max_speed(0.2)
```
Comment (jgkamat) [inline soccer/gameplay/plays/restarts/our_free_kick.py]: x

File: soccer/modeling/BallFilter.cpp [hunk 4]
```diff
@@ -30,9 +30,34 @@ Ball BallFilter::predict(RJ::Time time, float* velocityUncertainty) const {
...
     const auto s0 = vel.mag();
-    auto part = std::exp(-0.2913f * t.count());
-    auto speed = s0 * part;
-    auto distance = s0 * -3.43289f * (part - 1.0f);
+
+//    const auto decayConstant = 0.1735;
+    const auto decayConstant = 0.1795;
+
+
+//    auto part = std::exp(-0.2913f * t.count());
```
Comment (jgkamat) [inline soccer/modeling/BallFilter.cpp]: This part is copy pasted from https://github.com/RoboJackets/robocup-software/pull/911/files#diff-0050a07f81c01b7bc38ac572f8680d5dR88. 

Can you break it out into a function, perhaps?

File: soccer/planning/PivotPathPlanner.cpp [hunk 5]
```diff
@@ -89,23 +89,28 @@ std::unique_ptr<Path> PivotPathPlanner::run(PlanRequest& planRequest) {
...
-                    return AngleInstant(angleToPivotTarget);
-                } else {
-                    return AngleInstant(angleToPivot);
-                }
-            };
-        return make_unique<AngleFunctionPath>(move(path), function);
-        ;
+
+        if (path) {
+            // for (auto &waypoint: path->waypoints) {
```
Comment (jgkamat) [inline soccer/planning/PivotPathPlanner.cpp]: Can these be removed plese?

File: soccer/gameplay/plays/restarts/our_free_kick.py (hunk 3 above)
Comment (ashaw596) [reply soccer/gameplay/plays/restarts/our_free_kick.py]: ok.

File: soccer/modeling/BallFilter.cpp (hunk 4 above)
Comment (ashaw596) [reply soccer/modeling/BallFilter.cpp]: Whoops. Yeah I just used the Other ball filter.

**Comment by ashaw596** [comment]
Ok. That should be most the requested changes. Note the ball filter is tuned for real robots at competition.

File: soccer/gameplay/tactics/coordinated_pass.py [hunk 6]
```diff
@@ -146,11 +146,10 @@ def on_enter_preparing(self):
         kicker = self.skillkicker[0]
         kicker.target = self.receive_point
         kickpower = (main.ball().pos - self.receive_point).mag() / 8
-        if (kickpower < 0.2):
-            kickpower = 0.2
 
-        if (kickpower > 1.0):
-            kickpower = 1.0
+        kickpower = max(0.4, kickpower)
```
[the remark this answers is not in the document]
Comment (ashaw596) [reply soccer/gameplay/tactics/coordinated_pass.py]: ok

**Comment by jgkamat** [comment]
I'm going to merge this soon if we have no objections, I want to test this on real robots before competition this time :)

Just a heads up, @ashaw596, @jpfeltracco, @chachmu

**Pull request merged**